### ISSUE TITLE
FutureWarning from clip_grad_norm_ when training model in Python

### ISSUE THREAD

**Issue by Illumination720** [ultralytics/ultralytics#687]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am trying to train model from scratch in Python and got a warning like following:

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
  0%|          | 0/8 [00:00<?, ?it/s]C:\Users\lucas\Desktop\v8\ultralytics\yolo\engine\trainer.py:419: FutureWarning: Non-finite norm encountered in torch.nn.utils.clip_grad_norm_; continuing anyway. Note that the default behavior will change in a future release to error out if a non-finite total norm is encountered. At that point, setting error_if_nonfinite=false will be required to retain the old behavior.
  torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)  # clip gradients
       1/20       2.8G      1.264      1.286      1.281        225        640:  12%|█▎        | 1/8 [00:08<01:01,  8.75C:\Users\lucas\Desktop\v8\ultralytics\yolo\engine\trainer.py:419: FutureWarning: Non-finite norm encountered in torch.nn.utils.clip_grad_norm_; continuing anyway. Note that the default behavior will change in a future release to error out if a non-finite total norm is encountered. At that point, setting error_if_nonfinite=false will be required to retain the old behavior.
  torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)  # clip gradients
       1/20      2.81G      1.237      1.454      1.311        127        640:  25%|██▌       | 2/8 [00:10<00:26,  4.45C:\Users\lucas\Desktop\v8\ultralytics\yolo\engine\trainer.py:419: FutureWarning: Non-finite norm encountered in torch.nn.utils.clip_grad_norm_; continuing anyway. Note that the default behavior will change in a future release to error out if a non-finite total norm is encountered. At that point, setting error_if_nonfinite=false will be required to retain the old behavior.
  torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)  # clip gradients
       1/20      4.07G      1.174      1.435      1.252        217        640: 100%|██████████| 8/8 [00:17<00:00,  2.16
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?

The warning vanished after some time (Generally within the first epoch).

-------------------------------------------------------------------------------------------------------------------------------------

The reproduce code is like following:

```
import ultralytics
from ultralytics import YOLO

if __name__ == "__main__":
    model = YOLO("yolov8n.yaml")
    model.train(data='coco128.yaml',epochs=20,batch=16,workers=1)
```

or:
```
if __name__ == "__main__":
    model = YOLO("yolov8n.pt")
    model.train(data='coco128.yaml',epochs=20,batch=16,workers=1)
```

I wonder how this can be removed (not using warning filter) and does this have any impact to training or not. This warnings appear even pre-trained model is loaded. Not sure if this siutation is supposed to be like this and like to hear about suggestions and reasons for this happening.

Tried to set the learning rate lr0=0 but the warning still occured.

### Additional

Environment:
Windows 11

Pytorch version: 1.9.0

--------------------------------------------------------------------------------------------

The function produced the warning was located in:
ultralytics\yolo\engine\trainer.py:419

and inside the torch.nn.utils.clip_grad_norm_:

```
if total_norm.isnan() or total_norm.isinf():
        if error_if_nonfinite:
            raise RuntimeError(
                f'The total norm of order {norm_type} for gradients from '
                '`parameters` is non-finite, so it cannot be clipped. To disable '
                'this error and scale the gradients by the non-finite norm anyway, '
                'set `error_if_nonfinite=False`')
        else:
            warnings.warn("Non-finite norm encountered in torch.nn.utils.clip_grad_norm_; continuing anyway. "
                          "Note that the default behavior will change in a future release to error out "
                          "if a non-finite total norm is encountered. At that point, setting "
                          "error_if_nonfinite=false will be required to retain the old behavior.",
                          FutureWarning, stacklevel=2)
    clip_coef = max_norm / (total_norm + 1e-6)
    if clip_coef < 1:
        for p in parameters:
            p.grad.detach().mul_(clip_coef.to(p.grad.device))
    return total_norm
```

It suggests that total_norm has a value of inf. Want to check if this is supposed to be like this.

The dataset used is coco128.

**Comment by Laughing-q** [comment]
@Illumination720 I don't see any warnings by using the same command as yours, probably it's a environment issue.
![image](https://user-images.githubusercontent.com/61612323/215376411-36b0a114-432e-4304-9d28-c32a53b98c93.png)

**Comment by Illumination720** [comment]
> @Illumination720 I don't see any warnings by using the same command as yours, probably it's a environment issue. ![image](https://user-images.githubusercontent.com/61612323/215376411-36b0a114-432e-4304-9d28-c32a53b98c93.png)

Probably, the warning should be genrated when the version of torch is 1.9.0 as some values of self.model.parameters() is inf. The function clip_grad_norm_ throws the warning. I am confused if this is something supposed to be like this (inf value in parameters). BTW, is your torch version 1.9.0 (cuda)? I tested with train.py and run it by using **python train.py**

**Comment by Laughing-q** [comment]
@Illumination720 my torch version is 1.10.2. also using `python train.py`.

**Comment by Illumination720** [comment]
Yeah, the warning shouldn't apear under this version. In clip_grad_norm_ function:

torch 1.10
if error_if_nonfinite and torch.logical_or(total_norm.isnan(), total_norm.isinf()):
        raise RuntimeError(
            f'The total norm of order {norm_type} for gradients from '
            '`parameters` is non-finite, so it cannot be clipped. To disable '
            'this error and scale the gradients by the non-finite norm anyway, '
            'set `error_if_nonfinite=False`')
    clip_coef = max_norm / (total_norm + 1e-6)

torch 1.9.0
if total_norm.isnan() or total_norm.isinf():
        if error_if_nonfinite:
            raise RuntimeError(
                f'The total norm of order {norm_type} for gradients from '
                '`parameters` is non-finite, so it cannot be clipped. To disable '
                'this error and scale the gradients by the non-finite norm anyway, '
                'set `error_if_nonfinite=False`')
        else:
            warnings.warn("Non-finite norm encountered in torch.nn.utils.clip_grad_norm_; continuing anyway. "
                          "Note that the default behavior will change in a future release to error out "
                          "if a non-finite total norm is encountered. At that point, setting "
                          "error_if_nonfinite=false will be required to retain the old behavior.",
                          FutureWarning, stacklevel=2)
    clip_coef = max_norm / (total_norm + 1e-6)

I tested yolov8 yolo/engine/trainer part (line 418):
     self.scaler.unscale_(self.optimizer)  # unscale gradients
     torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)  # clip gradients
Some values in self.model.parameters() is inf and leads to the warning above. If this is okay in version 1.10.2. I would say the inf value is something normal in the first epoch while training.

**Comment by Illumination720** [comment]
Yeah, the warning shouldn't apear under this version. In clip_grad_norm_ function:

torch 1.10
if error_if_nonfinite and torch.logical_or(total_norm.isnan(), total_norm.isinf()):
        raise RuntimeError(
            f'The total norm of order {norm_type} for gradients from '
            '`parameters` is non-finite, so it cannot be clipped. To disable '
            'this error and scale the gradients by the non-finite norm anyway, '
            'set `error_if_nonfinite=False`')
    clip_coef = max_norm / (total_norm + 1e-6)

torch 1.9.0
if total_norm.isnan() or total_norm.isinf():
        if error_if_nonfinite:
            raise RuntimeError(
                f'The total norm of order {norm_type} for gradients from '
                '`parameters` is non-finite, so it cannot be clipped. To disable '
                'this error and scale the gradients by the non-finite norm anyway, '
                'set `error_if_nonfinite=False`')
        else:
            warnings.warn("Non-finite norm encountered in torch.nn.utils.clip_grad_norm_; continuing anyway. "
                          "Note that the default behavior will change in a future release to error out "
                          "if a non-finite total norm is encountered. At that point, setting "
                          "error_if_nonfinite=false will be required to retain the old behavior.",
                          FutureWarning, stacklevel=2)
    clip_coef = max_norm / (total_norm + 1e-6)

I tested yolov8 yolo/engine/trainer part (line 418):
     self.scaler.unscale_(self.optimizer)  # unscale gradients
     torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)  # clip gradients
Some values in self.model.parameters() is inf and leads to the warning above. If this is okay in version 1.10.2. I would say the inf value is something normal in the first epoch while training.

**Comment by pderrenger** [comment]
@Illumination720 The warning you mentioned seems due to the parameter values. As per the changes introduced in torch 1.10, the code does not raise the warning message for infinite or NaN values. It looks like the behavior is deemed normal during training in the first epoch.
